<launch>
  <arg name="fleet_name" default="fleet"/>
  <arg name="intruding_uav_name" default="intruder"/>
  <arg name="net_size" default="6.0"/>
  <arg name="net_size_half" default="3.0"/>

  <arg name="fleet_x_P" default="3.0"/>
  <arg name="fleet_x_I" default="0.06"/>
  <arg name="fleet_x_D" default="2.5"/>

  <arg name="fleet_y_P" default="5.0"/>
  <arg name="fleet_y_I" default="0.08"/>
  <arg name="fleet_y_D" default="3.5"/>

  <arg name="fleet_z_P" default="3.0"/>
  <arg name="fleet_z_I" default="0.05"/>
  <arg name="fleet_z_D" default="2.5"/>


  <!-- Set Up Fleet (net and 4 drones) -->
  <include file="$(find anti_uav)/launch/single_mav.launch">
    <arg name="x" value="$(arg net_size_half)"/>
    <arg name="y" value="-$(arg net_size_half)"/>
    <arg name="z" value="0.1"/>
    <arg name="yaw" value="0"/>
    <arg name="mav_name" value="$(arg fleet_name)_uav1"/>
    <arg name="color" value="Blue"/>
    <arg name="path" value="1"/>
    <arg name="starting_node_id" value="1"/>
  </include>

  <include file="$(find anti_uav)/launch/single_mav.launch">
    <arg name="x" value="-$(arg net_size_half)"/>
    <arg name="y" value="-$(arg net_size_half)"/>
    <arg name="z" value="0.1"/>
    <arg name="yaw" value="0"/>
    <arg name="mav_name" value="$(arg fleet_name)_uav2"/>
    <arg name="color" value="Blue"/>
    <arg name="path" value="1"/>
    <arg name="starting_node_id" value="2"/>
  </include>

  <include file="$(find anti_uav)/launch/single_mav.launch">
    <arg name="x" value="-$(arg net_size_half)"/>
    <arg name="y" value="$(arg net_size_half)"/>
    <arg name="z" value="0.1"/>
    <arg name="yaw" value="0"/>
    <arg name="mav_name" value="$(arg fleet_name)_uav3"/>
    <arg name="color" value="Blue"/>
    <arg name="path" value="1"/>
    <arg name="starting_node_id" value="3"/>
  </include>

  <include file="$(find anti_uav)/launch/single_mav.launch">
    <arg name="x" value="$(arg net_size_half)"/>
    <arg name="y" value="$(arg net_size_half)"/>
    <arg name="z" value="0.1"/>
    <arg name="yaw" value="0"/>
    <arg name="mav_name" value="$(arg fleet_name)_uav4"/>
    <arg name="color" value="Blue"/>
    <arg name="path" value="1"/>
    <arg name="starting_node_id" value="4"/>
  </include>


  <!-- Set Up Intruding UAV -->
  <include file="$(find anti_uav)/launch/single_mav.launch">
    <arg name="x" value="130.0"/>
    <arg name="y" value="0.0"/>
    <arg name="z" value="35.0"/>
    <arg name="mav_name" value="$(arg intruding_uav_name)"/>
    <arg name="color" value="Black"/>
  </include>


  <!-- UAV Position Controllers (for square vertices) -->
  <node name="uav1_controller" pkg="ros_copter" type="controller" output="screen">
    <param name="model_number" value="1"/>
    <param name="always_flying" value="true"/>

    <param name="u/P" value="0.0"/>
    <param name="u/I" value="0.0"/>
    <param name="u/D" value="0.0"/>

    <param name="v/P" value="0.0"/>
    <param name="v/I" value="0.0"/>
    <param name="v/D" value="0.0"/>

    <param name="x_P" value="$(arg fleet_x_P)"/>
    <param name="x_I" value="$(arg fleet_x_I)"/>
    <param name="x_D" value="$(arg fleet_x_D)"/>

    <param name="y_P" value="$(arg fleet_y_P)"/>
    <param name="y_I" value="$(arg fleet_y_I)"/>
    <param name="y_D" value="$(arg fleet_y_D)"/>

    <param name="z_P" value="$(arg fleet_z_P)"/>
    <param name="z_I" value="$(arg fleet_z_I)"/>
    <param name="z_D" value="$(arg fleet_z_D)"/>

    <param name="psi/P" value="0.0"/>
    <param name="psi/I" value="0.0"/>
    <param name="psi/D" value="0.0"/>

    <param name="max_throttle" value="1.0"/>

    <remap from="estimate" to="$(arg fleet_name)/uav1/ground_truth/odometry"/>
    <remap from="extended_command" to="$(arg fleet_name)/uav1/extended_command"/>
    <remap from="waypoint" to="uav1_goal"/>
  </node>

  <node name="uav2_controller" pkg="ros_copter" type="controller">
    <param name="model_number" value="2"/>
    <param name="always_flying" value="true"/>

    <param name="u/P" value="0.0"/>
    <param name="u/I" value="0.0"/>
    <param name="u/D" value="0.0"/>

    <param name="v/P" value="0.0"/>
    <param name="v/I" value="0.0"/>
    <param name="v/D" value="0.0"/>

    <param name="x_P" value="$(arg fleet_x_P)"/>
    <param name="x_I" value="$(arg fleet_x_I)"/>
    <param name="x_D" value="$(arg fleet_x_D)"/>

    <param name="y_P" value="$(arg fleet_y_P)"/>
    <param name="y_I" value="$(arg fleet_y_I)"/>
    <param name="y_D" value="$(arg fleet_y_D)"/>

    <param name="z_P" value="$(arg fleet_z_P)"/>
    <param name="z_I" value="$(arg fleet_z_I)"/>
    <param name="z_D" value="$(arg fleet_z_D)"/>

    <param name="psi/P" value="0.0"/>
    <param name="psi/I" value="0.0"/>
    <param name="psi/D" value="0.0"/>

    <param name="max_throttle" value="1.0"/>

    <remap from="estimate" to="$(arg fleet_name)/uav2/ground_truth/odometry"/>
    <remap from="extended_command" to="$(arg fleet_name)/uav2/extended_command"/>
    <remap from="waypoint" to="uav2_goal"/>
  </node>

  <node name="uav3_controller" pkg="ros_copter" type="controller">
    <param name="model_number" value="3"/>
    <param name="always_flying" value="true"/>

    <param name="u/P" value="0.0"/>
    <param name="u/P" value="0.0"/>
    <param name="u/I" value="0.0"/>
    <param name="u/D" value="0.0"/>

    <param name="v/P" value="0.0"/>
    <param name="v/I" value="0.0"/>
    <param name="v/D" value="0.0"/>

    <param name="x_P" value="$(arg fleet_x_P)"/>
    <param name="x_I" value="$(arg fleet_x_I)"/>
    <param name="x_D" value="$(arg fleet_x_D)"/>

    <param name="y_P" value="$(arg fleet_y_P)"/>
    <param name="y_I" value="$(arg fleet_y_I)"/>
    <param name="y_D" value="$(arg fleet_y_D)"/>

    <param name="z_P" value="$(arg fleet_z_P)"/>
    <param name="z_I" value="$(arg fleet_z_I)"/>
    <param name="z_D" value="$(arg fleet_z_D)"/>

    <param name="psi/P" value="0.0"/>
    <param name="psi/I" value="0.0"/>
    <param name="psi/D" value="0.0"/>

    <param name="max_throttle" value="1.0"/>

    <remap from="estimate" to="$(arg fleet_name)/uav3/ground_truth/odometry"/>
    <remap from="extended_command" to="$(arg fleet_name)/uav3/extended_command"/>
    <remap from="waypoint" to="uav3_goal"/>
  </node>

  <node name="uav4_controller" pkg="ros_copter" type="controller">
    <param name="model_number" value="4"/>
    <param name="always_flying" value="true"/>

    <param name="u/P" value="0.0"/>
    <param name="u/P" value="0.0"/>
    <param name="u/I" value="0.0"/>
    <param name="u/D" value="0.0"/>

    <param name="v/P" value="0.0"/>
    <param name="v/I" value="0.0"/>
    <param name="v/D" value="0.0"/>

    <param name="x_P" value="$(arg fleet_x_P)"/>
    <param name="x_I" value="$(arg fleet_x_I)"/>
    <param name="x_D" value="$(arg fleet_x_D)"/>

    <param name="y_P" value="$(arg fleet_y_P)"/>
    <param name="y_I" value="$(arg fleet_y_I)"/>
    <param name="y_D" value="$(arg fleet_y_D)"/>

    <param name="z_P" value="$(arg fleet_z_P)"/>
    <param name="z_I" value="$(arg fleet_z_I)"/>
    <param name="z_D" value="$(arg fleet_z_D)"/>

    <param name="psi/P" value="0.0"/>
    <param name="psi/I" value="0.0"/>
    <param name="psi/D" value="0.0"/>

    <param name="max_throttle" value="1.0"/>

    <remap from="estimate" to="$(arg fleet_name)/uav4/ground_truth/odometry"/>
    <remap from="extended_command" to="$(arg fleet_name)/uav4/extended_command"/>
    <remap from="waypoint" to="uav4_goal"/>
  </node>

  <!-- Intruder Position Controller -->
  <node name="controller" pkg="ros_copter" type="controller">
    <param name="u/P" value="0.0"/>
    <param name="u/I" value="0.0"/>
    <param name="u/D" value="0.0"/>

    <param name="v/P" value="0.0"/>
    <param name="v/I" value="0.0"/>
    <param name="v/D" value="0.0"/>

    <param name="x/P" value="0.0"/>
    <param name="x/I" value="0.0"/>
    <param name="x/D" value="0.0"/>

    <param name="y/P" value="0.0"/>
    <param name="y/I" value="0.0"/>
    <param name="y/D" value="0.0"/>

    <param name="z_P" value="3.0"/>
    <param name="z_I" value="0.05"/>
    <param name="z_D" value="2.5"/>

    <param name="psi/P" value="0.0"/>
    <param name="psi/I" value="0.0"/>
    <param name="psi/D" value="0.0"/>

    <param name="max_throttle" value="1.0"/>
    <param name="max_u" value="20.0"/>
    <param name="max_v" value="20.0"/>

    <remap from="estimate" to="$(arg intruding_uav_name)/ground_truth/odometry"/>
    <remap from="extended_command" to="$(arg intruding_uav_name)/extended_command"/>
  </node>

</launch>
